ksqldb-engine/src/main/java/io/confluent/ksql/query/KafkaStreamsQueryValidator.java

```diff
@@ -69,10 +70,32 @@ private void validateCacheBytesUsage(
       return;
     }
     final long configured = getCacheMaxBytesBuffering(config);
-    final long usedByRunning = running.stream()
-        .mapToLong(r -> new StreamsConfig(r.getStreamsProperties())
-                .getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG))
-        .sum();
+    long usedByRunning;
+    if (!config.getConfig(true).getBoolean(KsqlConfig.KSQL_SHARED_RUNTIME_ENABLED)) {
+      usedByRunning = running.stream()
+          .mapToLong(r -> new StreamsConfig(r.getStreamsProperties())
+              .getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG))
+          .sum();
+    } else {
+      usedByRunning = running.stream()
+          .filter(t -> !(t instanceof BinPackedPersistentQueryMetadataImpl))
+          .mapToLong(r -> new StreamsConfig(r.getStreamsProperties())
+              .getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG))
+          .sum();
+      usedByRunning += running.stream()
+          .filter(t -> t instanceof BinPackedPersistentQueryMetadataImpl)
+          .collect(HashMap<String, Long>::new,
+              (r, e) ->
+                  r.put(
+                      e.getQueryApplicationId(),
+                      (Long) e.getStreamsProperties()
+                          .get(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG)),
+              (r, l) -> l.putAll(r))
+          .values()
+          .stream()
+          .mapToLong(t -> t)
+          .sum();
+    }
```

Comment:
Can we just use a single `for` loop here? This is kind of hard to follow and the double iteration is unnecessary (idk how smart the java compiler is, maybe it would optimize this but maybe not)

Reply:
Should we also maybe validate that every runtime has the same value? Maybe just log a warning if there's a discrepancy don't throw an error or anything, for one thing it's not that severe and for a second I wonder if there might be a race condition if the user is trying to modify the system configs at the same time